third_party/move/move-model/src/builder/exp_builder.rs

```diff
@@ -3001,45 +3008,61 @@ impl<'env, 'translator, 'module_translator> ExpTranslator<'env, 'translator, 'mo
     /// Adds conversions to the given arguments for the given resolved function entry. Currently
     /// the only supported conversion is from `&mut T` to `&T` and we treat with it in an ad-hoc
     /// manor.
-    fn add_conversions(&self, entry: &AnyFunEntry, instantiation: &[Type], args: &mut [Exp]) {
+    fn add_conversions(
+        &self,
+        entry: &AnyFunEntry,
+        instantiation: &[Type],
+        args: Vec<Exp>,
+    ) -> Vec<Exp> {
+        let mut new_args = vec![];
```

Comment:
We can again use a functional style.  Let's see if it is any clearer:
```
{
    let params = entry.get_signature().1;
    let new_args = params.iter().map(|Parameter(_, ty, _)| ty.instantiate(instantiation))
        .zip(args)
        .map(|(param_ty, exp)| {
                    let exp_ty = self.env().get_node_type()(exp.node_id()),
                    self.try_freeze(&param_ty, &exp_ty, exp)
                  })
        .collect_vec();
    new_args
}
```
Seems nicer to me.